### ISSUE TITLE
Instructions for building for Linux

### ISSUE THREAD

**Issue by shuvendu-lahiri** [microsoft/verisol#96]
@rongpan @Yuepeng-Wang  can one of you document the steps needed in the Wiki to run the regression on a Linux box?

**Comment by garbervetsky** [comment]
I can try to do it for OSX but I couldn't make concurrency explorer work.

- Sent from my phone

**Comment by Yuepeng-Wang** [comment]
@shuvendu-lahiri I'll document the steps needed for Linux. Does the Wiki mean the file INSTALL.md?

**Comment by Yuepeng-Wang** [comment]
@shuvendu-lahiri Never mind. I find the Wiki pages now.

**Comment by Yuepeng-Wang** [comment]
Finished the first version at https://github.com/Microsoft/verisol/wiki/How-to-run-regressions-in-Linux

**Comment by shuvendu-lahiri** [comment]
@garbervetsky thanks...for the regressions, we don't need concurrencyExplorer. 
@Yuepeng-Wang thanks! I added a note at the end of INSTALL.md to your page. For building Corral, do you follow the instructions for "Building and Running Corral on Linux using Mono"? What about Z3? Do you not need Z3 as well? Also, are you able to get ConcurrencyExplorer on Linux?

**Comment by Yuepeng-Wang** [comment]
Yes, I follow the instructions "Building and Running Corral on Linux using Mono" for building Corral. It also needs Z3 as the underlying SMT solver. I'm not able to get ConcurrencyExplorer on Linux currently.

**Comment by garbervetsky** [comment]
In the process of making the instructions for OSX I tried to build Corral from scratch and get a couple of errors in Corral: 
	BoogieVerify.cs(317,40): error CS0176: Member 'StratifiedVCGen.callTree' cannot be accessed with an instance reference; qualify it with a type name instead
	BoogieVerify.cs(318,21): error CS0176: Member 'StratifiedVCGen.callTree' cannot be accessed with an instance reference; qualify it with a type name instead

When I see the file [StratifiedVC, line 834](https://github.com/boogie-org/boogie/blob/6ecd6bba5339f018a8b972a3576ced9a1c19c79d/Source/VCGeneration/StratifiedVC.cs#L834) in Boogie I found the variable is static: `public static HashSet<string> callTree = null;`

I see that the [this commit](https://github.com/boogie-org/corral/commit/88b7f62929452755a004d29868d833d1ebad8c8e) is producing this error.

**Comment by garbervetsky** [comment]
I found some issues to run `SolToBoogieTest.dll` in my mac (a missing netcore library). I'll leave a draft instructions but I need to see how to fix this issue.

**Comment by garbervetsky** [comment]
> I found some issues to run `SolToBoogieTest.dll` in my mac (a missing netcore library). I'll leave a draft instructions but I need to see how to fix this issue.

Update: I try a clean install in my other laptop and works fine! 👍

**Comment by shuvendu-lahiri** [comment]
Thanks @garbervetsky  @Yuepeng-Wang 

However, I don't see instructions to install Z3 for linux/OSX. Is it present in the Corral Linux instructions?

**Comment by garbervetsky** [comment]
good point.

in my case it is just creating a symbolic link (z3.exe) in corral binaries.
I'll update the instructions. I assumed that is included in the
instructions for corral...

thanks.

- Sent from my phone

**Comment by garbervetsky** [comment]
As I mentioned,  the instructions about [how to install corral for Linux](https://github.com/boogie-org/corral#building-and-running-corral-on-linux-using-mono) already explain how to link `z3`. 
I think we don't need to add additional information.

**Comment by Yuepeng-Wang** [comment]
Yes, the instructions for building Corral on Linux describe how to link Z3 for Corral. I changed the Wiki to refer to the instructions.

**Comment by VeraBE** [comment]
I think there's an error in the Linux instructions, the command to run the regressions is `dotnet %VERISOL_PATH%\Sources\SolToBoogieTest\bin\Debug\netcoreapp2.2\SolToBoogieTest.dll %VERISOL_PATH%\test\` not `dotnet $VeriSolPath/Sources/SolToBoogieTest/bin/Debug/netcoreapp2.0/SolToBoogieTest.dll <Path To Corral Executable> $VeriSolPath`.

I also had to copy `solc-static-linux` manually to `SolToBoogie\bin\Debug\netcoreapp2.2\` and run `chmod a+x solc-static-linux`, because the commands in `SolToBoogieTest.csproj`, `VeriSolOutOfBandSpecsRunner.csproj` and `SolToBoogie.csproj` use the Windows' `copy` command and `solc.exe`. That error made the build command show "Build FAILED".

For the tests to pass I also had to change line 62 in `SolToBoogieTest/Program.cs` because the corral folder should be lowercase, the debug folder should be uppercase and corral.exe lowercase too.

**Comment by Yuepeng-Wang** [comment]
@VeraBE Thanks for pointing out the problem! I've updated the instructions about execution permissions.

I don't have any path related issues on Ubuntu 18.04. Are you using Cygwin or running Linux on a virtual machine? Otherwise, the paths in Linux should use slash (/).

**Comment by VeraBE** [comment]
> I don't have any path related issues on Ubuntu 18.04. Are you using Cygwin or running Linux on a virtual machine? Otherwise, the paths in Linux should use slash (/).

You're referring to the first thing that I mentioned? Sorry if I used the wrong slashes in the comment, I was trying to ask if the parameter of the DLL should be the test directory instead of the Corral executable and VeriSol directory as the instructions say.

**Comment by shuvendu-lahiri** [comment]
@VeraBE I have recently updated some changes (e.g. making Corral as a submodule), and changed some of the instructions for Windows. Hence I asked @garbervetsky to update the Linux instructions. You rightly point out the change needed to the regression script.

**Comment by shuvendu-lahiri** [comment]
@Yuepeng-Wang sorry for catching you off guard on this issue!

**Comment by shuvendu-lahiri** [comment]
Please see https://github.com/Microsoft/verisol/blob/master/INSTALL.md#regression-script for updated windows instructions. I will update the instructions later, unless one of you do.

**Comment by shuvendu-lahiri** [comment]
Also, given Corral is a submodule, there is no need to download corral. Can you add the instructions to build the verisol\Corral\ folder in the instruction page? Look at the windows instructions for reference. @Yuepeng-Wang since you are updating the instructions.

**Comment by Yuepeng-Wang** [comment]
@shuvendu-lahiri  I've updated the Linux instructions for using Corral as a git submodule.
But I also get "Build Failed" because of line 17 in
```
Sources/SolToBoogie/SolToBoogie.csproj
Sources/SolToBoogieTest/SolToBoogieTest.csproj
Sources/VeriSolRunner/VeriSolOutOfBandSpecsRunner.csproj
```
Is there a way to copy the file that works on all platforms?

**Comment by garbervetsky** [comment]
I believe one solution would be removing the line that does "copy solc.exe"
from those projects.

**Comment by shuvendu-lahiri** [comment]
Fixed for now by removing the "copy" #120

**Comment by VeraBE** [comment]
> For the tests to pass I also had to change line 62 in `SolToBoogieTest/Program.cs` and rebuild it because the corral folder should be lowercase, the debug folder should be uppercase and corral.exe lowercase too.

This is still needed
